### DISCUSSION TITLE
Possibility of a website scraping using the internal browser and JS?

### DISCUSSION THREAD

**Discussion by DawEdhel** [martinrotter/rssguard#1313] [Q&A] [answered]
My main problem is the data I want to retrieve is hidden behind an authorization process and the page is powered by React framework.
I've already tried Puppeteer, but to no avail.

Is it possible to do the following steps or in any similar way?
1. Login in advance to a target URL using the internal browser;
2. Provide the URL to the feed's source field and force the app to retrieve the DOM source after a "idle network" event via the internal browser using the existing cookies;
3. Provide a JS script to the feed's postscript field to execute on the DOM source and pass the result of the script (e.g. JSON) to the app to parse;

**Comment by martinrotter** [comment]
This actually doable idea. Will check it out but should be possible via QWebEnginePage

**Comment by martinrotter** [comment]
OK, I pushed the code and development build is now compilling. Please test when compiled and report back.

1. I added new source type "Built-in web browser". Simply select that and put in your URL.
2. Make sure to be "logged-in" that URL in built-in browser before step 1.

Here is how the dialog setup looks (do not mind the third red rectangle, that is only for my purposes for that specific testing website I used).
![image](https://github.com/martinrotter/rssguard/assets/1255302/30630a2d-057b-4353-9cd3-4fdceb5b1849)

3. The feature works in both "Fetch it now" testing mode and real live mode.
4. The testing website I use provides the HTML only when user is logged-in and when fetched with this new feature it correctly returns expected data (which I additionally parse with "css2rss").

![image](https://github.com/martinrotter/rssguard/assets/1255302/3e92b644-6146-4f8b-b515-84a180c4d61c)

The post-process field yes. Read documentation. Simply when the data is obtained from the source with one of methods (URL, local file, Script or new Built-in web browser URL) then the data is fed to provided post-process script via console standard input/output mechanism).

You can likely use node to execute some JS on HTML data I guess you need to write custom Node script which:
1. Reads standard input (the obtained data).
2. Creates DOM from the input.
3. Runs the JS you need.
4. Prints the modified DOM to the standard output (which is then read back by RSS guard and parsed as a feed).

So your point "3" is entirely in your direction, all the tools necessary are already there in RSS Guard.

For point 2 - the part of retrieving the webpage via internal browser is implemented as per I wrote above. But, what exactly you mean by 
**after a "idle network" event via the internal browser**. You mean that internal web browser should wait until some specific JS event/message arrives? If so, how exactly this event looks like?

**Reply by DawEdhel** [reply]
> I added new source type "Built-in web browser". Simply select that and put in your URL.

I've tested the dev build, and can confirm the authorization part being done successfully.

Since there is no way in the app to quickly see the retrieved stdin data, I wrote a quick python script to save it to a local file. And after examining it, I didn't see the data I needed but rather an empty template that should gets populated later in the page cycle via fetch/xhr, which leads to your last question:
> But, what exactly you mean by after a "idle network" event via the internal browser. You mean that internal web browser should wait until some specific JS  event/message arrives? If so, how exactly this event looks like?

[See this Puppeteer reference for the "idle network" event.](https://pptr.dev/api/puppeteer.puppeteerlifecycleevent/)
This is how, I suppose, they deal with React pages: it waits for most initial network connections to be processed first (including async scripts after document.ready event). It could be very flexible to be used if both these amount of network connections and wait timeout values can be provided by an user as arguments to the "idle network" event.

If it's not achievable, I would like to get a 'wait for some time after the document is ready and then return the HTML source' timeout argument instead, to make it possible to be dealt with.

**Reply by martinrotter** [reply]
Can you provide me some html file with embedded js for me to test the feature and allies to see that js was executed properly to tweak the html?

**Reply by DawEdhel** [reply]
No, it's somewhat private.
But, if you wish, I can try to replicate the problem with a public website and provide selectors to test with.

**Reply by martinrotter** [reply]
I will try to hack something Maybe simply load youtube site will just work?

**Reply by DawEdhel** [reply]
I've prepared an example based on main page of [airbnb](https://www.airbnb.com/):

Items' selector: 'main#site-content div[style*="display:contents"] > div > div > div[class=" dir dir-ltr"] div[itemprop="itemListElement"] > div > div > div > a'

In a browser it will output 20 elements on load, while nothing via the dev build's stdout.
The main difference happens at this node: "div[class=" dir dir-ltr"] (same as in the selector above) > div", where the first div child is a template that visible in the stdout, but hidden in the browser upon finish state; and the second div child is empty in the stdout, but is populated in the browser and contains 'div[itemprop="itemListElement"]' element.

Browser's network tab show that DOM was prepared in ~1.5 sec, while the items were supplied in ~5 sec overall.

**Reply by martinrotter** [reply]
@DawEdhel 

OK. You airbnb example was not working for me because of that crazy CSS selector.

However, I made some changes which actually should allow proper JS execution in fake headless mode. The thing is that (some) JS is only executed if the website thinks that the page is actually displayed. Therefore I had to create "invisible" QWebEngineView instance to fake this. Note that this is memory hungry so repeated crazy feed fetching with this might not be that great.

Also, I added artificial 15 seconds delay which runs after the website DOM is loaded to give JS time to do its thing.

We will figure universal way of determining "when" the result should be grabbed later.

I tested with website https://www.modrastrecha.cz/ which has on the right side panel "Aktivní diskuse" which is loaded by JS with post-process script:

> python3 css2rss.py '#threads-widget a'

It gives correct result in "Fetch it now" test mode and in live feed fetching for me.

![image](https://github.com/martinrotter/rssguard/assets/1255302/252cd345-4285-492c-b1d4-8fd455cab1b0)

It seems that "some" (?) websites use some kind of JS events/mechanisms which actually need physical (?) screen to JS be evaluated and did not work for me so far, like Youtube. But will test that again.

Please do your testing, code should be compiled soon and report back.

Also note that maybe this hassle is overkill because there are solutions for headless JS-enabled sites fetching like pupeteer or hlspy. Also I myself use yet another python-based headless Chrome-based library to scrape one JS-enabled website for me.

```py
import undetected_chromedriver as uc
import sys
import psutil

sys.stdout.reconfigure(encoding="utf-8")

options = uc.ChromeOptions()
options.headless = True
options.add_argument("--headless")
options.binary_location = ".\\chrome\\chrome.exe"

chrome = uc.Chrome(options=options)

chrome.get("https://antivirus.22web.org/clanky.htm")
chrome.implicitly_wait(20)

print(chrome.page_source.replace('href="av/', 'href="https://antivirus.22web.org/av/'))
chrome.quit()

def terminate_crashpad_processes():
    target_type = ['--type=crashpad-handler', '--type=renderer']

    for process in psutil.process_iter(['pid', 'name', 'cmdline']):
        if process.info['name'] == 'chrome.exe' and any(tt in process.info['cmdline'] for tt in target_type):
            try:
                psutil.Process(process.info['pid']).terminate()
                print(f"Terminated (PID: {process.info['pid']})")
            except Exception as e:
                print(f"Error terminating (PID: {process.info['pid']}): {e}")

terminate_crashpad_processes()
```

That is the script I use in "source" field in RSS guard.

**Reply by martinrotter** [reply]
More tests (of course first I opened the website in RSS Guard browser new tab and accepted all cookies and sometimes logged-in):

reddit works
![image](https://github.com/martinrotter/rssguard/assets/1255302/5bb13a76-3a06-48e7-8b92-aaaf24bda439)

**Reply by martinrotter** [reply]
Progress: I wrote simple DOM change observer which waits for some time before it reports that DOM is "idle".

```js
// Watch entire document and keep date of last DOM change.
const targetNode = document;
const waitTimeMs = 5000;
var lastResourceTime = new Date();

// Setup DOM observer and observe for changes in elements only.
const config = { attributes: false, characterData: false, childList: true, subtree: true };

const callback = (mutationList, observer) => {
  lastResourceTime = new Date();
  console.log("res " + lastResourceTime);
};

const observer = new MutationObserver(callback);

observer.observe(targetNode, config);

var intervalId = window.setInterval(function(){
  var actualTime = new Date();

  if ((actualTime - lastResourceTime) > waitTimeMs) {
    console.log("llooaaddeedd");

    clearInterval(intervalId);
  }
  /*
  else {
    console.log("last " + lastResourceTime);
    console.log("act " + actualTime);
  }
  */
}, 2000);
```

**Reply by martinrotter** [reply]
OK, will push the code soon, how it works.

1. Initial DOM of the target website is loaded.
2. The above javascript is injected.
3. When JS detects "idle" state, it sends special string to "console.log"
4. RSS Guard detects the special string and then extracts HTML string of the DOM.
5. HTML string is returned for further processing.

**Reply by DawEdhel** [reply]
> It seems that "some" (?) websites use some kind of JS events/mechanisms which actually need physical (?) screen to JS be evaluated and did not work for me so far, like Youtube. But will test that again.
>Also note that maybe this hassle is overkill because there are solutions for headless JS-enabled sites fetching like pupeteer or hlspy.

That's why I made this discussion initially: either Puppeteer has browser's limited capabilities or websites use some scraping counter-measures against it, I'm not sure, but having the most genuine browser build, as what an average user would use, might help in that case, what you use here as an internal browser, I assume.

I made a note of your python-based solution, but I'd try it as a last resort.

> OK, will push the code soon, how it works.

I've downloaded the devbuild-51ddab938 and tried to retrieve the stdin through my python script in the post-process field, but to no avail: after pressing 'Fetch it now' button, the 'Add new feed' window went unresponsive for more than couple minutes and so on, until I force-closed the application; same goes to the 'Application log' window.

I put your observer script to be executed in the console of my private example at the start of it as soon as possible: the special string were returned in ~10 sec after the start of its execution, so the script wasn't in any sort of a constant loop (there should be a max wait timeout from the starting of the observer in case of websites constantly changing their DOM over long period of time).

P.S. After force-closing the app and opening again, it started complaining about Adblock being disabled. I've found there are some ghost leftovers of nodejs runtime processes, and after terminating them the app could start the extension successfully.

**Reply by martinrotter** [reply]
I just got an idea, will implement and we can then retest.

**Reply by martinrotter** [reply]
OK so some summary:

* [x] - Authentication, cookies, storage - this all seems to work very nicely, state is remembered, logins too.
* [x] - JavaScript itself runs without problems and the internal JS script I wrote is able to detect "idle" moments very well.
* [ ] - Problem is that some websites use JS techniques which run JavaScript only when page becomes physically visible/drawn. The solution to this is veeeeery problematic as the only way is literally showing the temporary "web browser" somewhere in the app which is just unacceptable for vast majority of users. Just tested this on Reddit subreddits. If I hack intermediate web browser and display it while the feed fetches the data from it, then all HTML elements are generated by JS properly just as we expect and I get all reddit data. But when the "web browser" is created but hidden, well, the JS magic simply does not run.

So, implementation of this feature will work for some JS-enabled sites, but not for all sites. Not sure what approach proper Chrome headless mode uses, if it works better or not.

I would likely recommend to stick to those solutions, @DawEdhel what you think.

**Reply by DawEdhel** [reply]
> Problem is that some websites use JS techniques which run JavaScript only when page becomes physically visible/drawn.

I've made some research about this, and my question is: have you tried dealing with Document.visibilityState property?
> The Document.visibilityState read-only property returns the visibility of the document. It can be used to check whether the document is in the background or in a minimized window, or is otherwise not visible to the user.

[Here](https://gist.github.com/gwarser/3b47b61863bffcfebe4498c77b2301cd#file-disable-pageview-api-js) is one of examples of its spoofing. There is also [this solution](https://sqa.stackexchange.com/a/32355).

——————

Overall, it's still a better solution over URL page source's option in some scenarios, so I'll take anything there is.

P.S. I'm still not sure what to do with my latest 'unresponsive result' testing.

**Reply by martinrotter** [reply, marked as answer]
@DawEdhel I pushed new code, cleaned the feature up and fixed some related bugs which could be responsible for your unresponsiveness.

I have also tested it with some websites and it indeed works. I also tweaked built in JS and now it automatically triggers 1 scrolling to bottom, which in some cases forces the page to load more of its JS-enabled content.

Please, test latest devbuild when it compiles, I will take a look on the spoofing thing you sent.

EDIT: With my limited JS knowledge I have tested both your suggestions - neither seems to work sadly - tested with Youtube.

**Reply by DawEdhel** [reply]
@martinrotter

I've downloaded latest devbuild-07eb444d9 and tested it, and indeed, I could successfully retrieve a populated with data HTML of my initially problematic example, which makes initial post of mine resolved.

But there were some caveats: when I wrote a script to parse HTML and provide a JSON feed, then pressed "Fetch it now" button, I got an "Error: feed format not recognized". Initially thought I wasn't following the feed's format, but couldn't find anything. Then I've tried to create a new feed, using a working-in-the-stable-build local file with JSON, and I got the same error.

I managed to circumvent this error by doing the following:
1. Supply into the "Add new feed" window any URL source with a RSS feed, press the "Fetch it now" button, then OK button.
2. Select it to edit, change all needed fields to expect JSON from the built-in browser and post-script, then OK button.
3. "Fetch selected" it, and articles list gets populated successfully with no errors.

Looks like "Fetch it now" and "Fetch selected" buttons follow different pathways for some reason in this dev build.

P.S. I would like to look into the Youtube problem. Which part of it couldn't be retrieved? Videos on its main page?

**Reply by martinrotter** [reply]
Maybe provide your setup first so we can find what was wrong when you was testing that JSON.

As for youtube yes, the page loaded, but video are simply not there

**Reply by DawEdhel** [reply]
Fresh unpacking of rssguard-devbuild-07eb444d9-win10.7z with no previous data (portable).

Inside the "Add new feed" window:
Type: JSON
Source: Local file with a path target
File's contents as an example: "{ "title": "Test", "items": [] }"

"Fetch it now" responds with an "Error: feed format not recognized".
"Application log" window's messages:
time="  1113.736" type="warning" -> core: Feed guessing error: 'XML is not well-formed, Start tag expected.'.
time="  1113.736" type="warning" -> core: Feed guessing error: 'XML is not well-formed, Start tag expected.'.
time="  1113.737" type="warning" -> core: Feed guessing error: 'XML is not well-formed, Start tag expected.'.
time="  1113.741" type="warning" -> core: Feed guessing error: 'not a JSON feed'.
time="  1113.742" type="warning" -> core: Feed guessing error: 'XML is not well-formed, Start tag expected.'.

**Reply by martinrotter** [reply]
That is not a valid JSON feed 1.1.

**Reply by DawEdhel** [reply]
I've figured out the JSON validation have become more strict with its format during the dev builds, and made an appropriate change to my script. Using it, I've created a new feed with built-in browser option, and fetched it afterwards to populate articles list, successfully.

Thus I'm going to mark this discussion resolved and await for the newer stable release for that feature. Thanks for adding it.

**Reply by DawEdhel** [reply]
> As for youtube yes, the page loaded, but video are simply not there

About this: I've tried nodejs/puppeteer to retrieve the source of site's main page as a headless browser, and it successfully returned a list of random videos pretty much with minimum efforts.

Here's the code:
```
const puppeteer = require('puppeteer');
const fs = require('fs').promises;

(async() => {
	
    const browser = await puppeteer.launch({
    	headless: true,
    });
    const page = await browser.newPage();
    
    // generate cookies in advance, by using a headful mode and clicking on videos/shorts, to appease video recommendations:
    // const cookies = await page.cookies();
    // await fs.writeFile('./youtube-cookies.json', JSON.stringify(cookies, null, 2));
    const cookiesString = await fs.readFile('./youtube-cookies.json');
    const cookies = JSON.parse(cookiesString);
    await page.setCookie(...cookies);
    
    await page.goto('https://www.youtube.com/', {
    	timeout: 30000,
    	waitUntil: 'networkidle2'
    });
    
    await page.screenshot({path: 'youtube.png'})
    await fs.writeFile('./youtube-source.txt', await page.content());
    
    await browser.close();
	
})();
```

Multiple results have shown:
- browser was never visible during the whole process;
- screenshots confirmed it wasn't an empty page but populated with random videos every time;
- same goes to the source text, where `<div id="contents" class="style-scope ytd-rich-grid-renderer">` contained multiple rows of videos.

**Reply by martinrotter** [reply]
Yes, like I said, specialized solutions like selenium/pupeteer (which is great, indeed) are likely better than this built-int effort.
